### PR TITLE
Disallow merging existing mapping field definitions in templates

### PR DESCRIPTION
**Pull request by dakrone**
This commit changes the merge strategy introduced in #55607 and #55982. Instead of overwriting these
fields, we now prevent them from being merged with an exception when a user attempts to
overwrite a field.

As part of this, a more robust validation has been added. The existing validation checked whether
templates (composable and component) were valid on their own, this new validation now checks that
the composite template (mappings/settings/aliases) is valid. This means that when a composable
template is added or updated, we confirm that it is valid with its component pieces. When a
component template is updated we ensure that all composable templates that make use of the component
template continue to be valid before allowing the component template to be updated.

This change also necessitated changes in the tests, however, I have left tests that exercise mapping
merging with nested object fields as `@AwaitsFix`, as we intend to change the behavior soon to allow
merging in a recursive-with-replacement fashion (see: #57393). I have added tests that check the new
disallowing behavior in the meantime.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 1]
```diff
@@ -910,6 +951,77 @@ private static Settings resolveSettings(Metadata metadata, ComposableIndexTempla
...
+
+            // use the provided values, otherwise just pick valid dummy values
+            int dummyPartitionSize = IndexMetadata.INDEX_ROUTING_PARTITION_SIZE_SETTING.get(resolvedSettings);
+            int dummyShards = resolvedSettings.getAsInt(IndexMetadata.SETTING_NUMBER_OF_SHARDS,
+                dummyPartitionSize == 1 ? 1 : dummyPartitionSize + 1);
+            int shardReplicas = resolvedSettings.getAsInt(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0);
+
+
+            //create index service for parsing and validating "mappings"
+            Settings dummySettings = Settings.builder()
```
[the remark this answers is not in the document]
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:981]: also maybe `dummySettings` is not the best name as it includes the actual `resolvedSettings`? maybe `finalIndexSettings` or `resolvedIndexSettings`? What do you think?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 2]
```diff
@@ -910,6 +951,77 @@ private static Settings resolveSettings(Metadata metadata, ComposableIndexTempla
...
+            IndexService dummyIndexService = indicesService.createIndex(tmpIndexMetadata, Collections.emptyList(), false);
+            createdIndex = dummyIndexService.index();
+
+            // Validate aliases
+            MetadataCreateIndexService.resolveAndValidateAliases(temporaryIndexName, Collections.emptySet(),
+                MetadataIndexTemplateService.resolveAliases(stateWithIndex.metadata(), templateName), stateWithIndex.metadata(),
+                new AliasValidator(),
+                // the context is only used for validation so it's fine to pass fake values for the
+                // shard id and the current timestamp
+                xContentRegistry, dummyIndexService.newQueryShardContext(0, null, () -> 0L, null));
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:1005]: would using `indicesService.withTempIndexService` be better as it avoids adding.removing the index from the `IndexService` and also handles closing the `indexService`? (we're using this in `TransportSimulateIndexTemplateAction`)

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateServiceTests.java [hunk 3]
```diff
@@ -859,6 +919,119 @@ public void testRemoveComponentTemplateInUse() throws Exception {
             containsString("component templates [ct] cannot be removed as they are still in use by index templates [template]"));
     }
 
+    /**
+     * Tests that we check that settings/mappings/etc are valid even after template composition,
+     * when adding/updating a composable index template
+     */
+    public void testInvalidComposableCompositeTemplate() throws Exception {
```
Comment (andreidan) [inline server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateServiceTests.java:926]: this and the next test names are rather confusing to me. It's quite difficult to pick up what they are testing and what's the invalid case. (maybe related to the previous comment too, but  "composable composite" adds to the confusion for me)

what do you think of including the cause of failure in the test name? 

What do you think of something along the lines of  `testIndexTemplateFailsToOverrideComponentTemplateMappingField` and respectively `testUpdateComponentTemplateFailsIfResolvedIndexTemplatesWouldBeInvalid` ?

**Review by andreidan — CHANGES_REQUESTED** [review]
Thanks for working on this Lee! I left a few mostly minor comments

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:981]: Yep, I think that makes sense, going with `finalResolvedSettings`

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java (hunk 2 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:1005]: Yes absolutely, I'll change this

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 4]
```diff
@@ -910,6 +951,77 @@ private static Settings resolveSettings(Metadata metadata, ComposableIndexTempla
         return Collections.unmodifiableList(aliases);
     }
 
+    /**
+     * Given a state and a composable template, validate that the final composite template
+     * generated by the composable template and all of its component templates contains valid
+     * settings, mappings, and aliases.
+     */
+    private static void validateCompositeTemplate(final ClusterState state,
```
[the remark this answers is not in the document]
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:959]: I dunno, I didn't think they were too different, I wanted to treat "resolved" as pertaining to a particular index, so I think the current name isn't too bad.

**Review by andreidan — APPROVED** [review]
LGTM, thanks for iterating on this Lee

**Comment by dakrone** [comment]
@elasticmachine update branch

**Comment by jtibshirani** [comment]
@dakrone and I discussed offline and determined that we should also throw an error when a metadata key is specified twice (like `_source: enabled`), and not just keys in the properties block. This will allow us to make changes to the metadata merge semantics in 7.9.

**Pull request merged**